### PR TITLE
feat: share short urls

### PR DESCRIPTION
**Pull request by rephus**
Closes:  #2225

### Description:

[Screencast from 10-10-22 11:06:55.webm](https://user-images.githubusercontent.com/1983672/194846236-b1c56aff-1f73-428f-a6ae-9ffd8f0b3c52.webm)

[Screencast from 10-10-22 13:29:03.webm](https://user-images.githubusercontent.com/1983672/194856586-39284dbd-d5a6-4237-bb69-55c2b42ce1f6.webm)

### CODE REVIEW

File: packages/backend/src/services/ShareService/ShareService.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+            url: `${shareUrl.path}${shareUrl.params}`,
+        };
+    }
+
+    async getShareUrl(nanoid: string): Promise<ShareUrl> {
+        const shareUrl = await this.shareModel.getSharedUrl(nanoid);
+        return this.shareUrlWithHost(shareUrl);
+    }
+
+    async createShareUrl(path: string, params: string): Promise<ShareUrl> {
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/services/ShareService/ShareService.ts]: Screenshot from 2022-10-11 14-09-25

File: packages/backend/src/database/entities/share.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { Knex } from 'knex';
+
+export const ShareTableName = 'share';
+
+export type DbShareUrl = {
+    nanoid: string;
+    path: string;
+    params: string;
```
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/database/entities/share.ts:8]: Screenshot from 2022-10-11 14-07-11

Screenshot from 2022-10-11 14-17-55

[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/backend/src/database/entities/share.ts:8]: Missed `created_at` in this type

File: packages/backend/src/services/ShareService/ShareService.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            url: `${shareUrl.path}${shareUrl.params}`,
+        };
+    }
+
+    async getShareUrl(user: SessionUser, nanoid: string): Promise<ShareUrl> {
+        if (user.ability.cannot('view', subject('Organization', user))) {
+            throw new ForbiddenError();
+        }
+
+        const shareUrl = await this.shareModel.getSharedUrl(nanoid);
```
Comment (ZeRego) [inline packages/backend/src/services/ShareService/ShareService.ts:38]: This permission check should check against the shareUrl org uuid

File: packages/backend/src/models/ShareModel.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            })
+            .returning('*');
+
+        return share;
+    }
+
+    async getSharedUrl(nanoid: string): Promise<ShareUrl> {
+        const [row] = await this.database(ShareTableName)
+            .where('nanoid', nanoid)
+            .select<ShareUrl[]>('*');
```
Comment (ZeRego) [inline packages/backend/src/models/ShareModel.ts:40]: We will have to join user and org tables to return the uuids.

**Review by ZeRego — CHANGES_REQUESTED** [review]

File: packages/backend/src/database/entities/share.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rephus) [reply packages/backend/src/database/entities/share.ts:8]: What do you mean ?

[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/backend/src/database/entities/share.ts:8]: We have a `created_at` column but is not part of the type  `DbShareUrl`

**Review by ZeRego — APPROVED** [review]
lgtm 🚀

**Comment by rephus** [comment]
Not sure what's wrong with render. retesting here https://github.com/lightdash/lightdash/pull/3466

**Comment by rephus** [comment]
Merged https://github.com/lightdash/lightdash/pull/3466

**Comment by rephus** [comment]
@12ian34 this is cool headway material

**Comment by 12ian34** [comment]
cc @hamzahc1

**Pull request closed without merging**